### DISCUSSION TITLE
Zookeeper와 broker 갯수의 대응 비율

### DISCUSSION THREAD

**Discussion by sejongjeong** [kyungjunleeme/kafka_study#9] [Q&A] [answered]
N:1이 효율적인지 1:1이 효율적인지?

**Comment by kyungjunleeme** [comment]
https://www.inflearn.com/community/questions/979478/%EC%A3%BC%ED%82%A4%ED%8D%BC%EC%9D%98-%EB%8C%80%ED%95%9C-%EC%97%AD%ED%95%A0%EC%9D%B4-%EC%9E%98-%EC%9D%B4%ED%95%B4%EA%B0%80-%EB%90%98%EC%A7%80-%EC%95%8A%EC%8A%B5%EB%8B%88%EB%8B%A4?srsltid=AfmBOoq5Uo3NqLyNLokOhVFWvOo25Nlonz8UrpMBo75HJgdk6gE_A9G3

주키퍼는 카프카 클러스터당 1개 앙상블이 필요합니다. 보통 주키퍼 앙상블은 상용환경에서 3개~5개 서버로 구성합니다.

**Comment by hellices** [comment, marked as answer]
주키퍼는 브로커 3대 중 리더를 지정하고, 각각의 브로커의 상태를 체크하도록 설계되어 있습니다.
이 때 주키퍼가 죽으면 브로커를 관리할 수 없으므로 주키퍼 또한 3대를 최소한의 운영 정책으로 가져가야 합니다.

이러한 구조를 쿼럼(Quorum)이라고 하며, 홀수의 구조로 안정성 및 데이터 일관성을 보장하는 전략입니다.
Quorum = ⌊N/2⌋ + 1

그러므로 운영 환경에서는 주키퍼와 브로커는 별도의 노드로 분리되어야 합니다.

주키퍼에 대한 의존도를 제거하기 위해서 카프카는 KRaft를 도입하고 있으며, 브로커 자체적으로 쿼럼 구조를 통해 자신들의 상태를 관리하는 구조로 진화하고 있습니다. 카프카 [문서](https://kafka.apache.org/documentation/#zk_depr)에 따르면 zookeeper에 의존하는 버전은 3.x에서 종료하고 향후 4.x버전이 출시될 경우 KRaft모드만 지원하는 것으로 나오고 있습니다.
